### PR TITLE
Declare the "celery" exchange as non-durable

### PR DESCRIPTION
**Pull request by seanh**
Fixes https://github.com/hypothesis/h-periodic/issues/118

AMQP is raising a `inequivalent arg 'durable' for exchange 'celery' in
vhost 'h': received 'true' but current is 'false'` meaning that the
`'celery'` exchange in CloudAMQP is set to `durable=False` (which is
correct) but h-periodic is trying to declare the exchange as
`durable=True` (which is wrong).

The fix is to change h-periodic to declare the exchange as
`durable=False`. This should make the errors go away.

There shouldn't be any functional change other than the error going
away. The error doesn't actually seem to mean that anything is broken,
it just seems to be warning.

We already do this for Checkmate:

https://github.com/hypothesis/h-periodic/blob/4de61bdd085eee923023a755eb2cca2385109ac9/h_periodic/checkmate_beat.py#L10-L18

We actually want the RabbitMQ queues, exchanges, etc used by h-periodic
to be non-durable. We don't need them to be durable for periodic tasks.
The way that periodic tasks work is that h-periodic constantly runs a
Celery beat process that adds one instance of each task to the queue
every minute (or whatever the time period is, according to the schedules
in the `h_periodic/*_beat.py` files). The task immediately gets consumed
by a worker. A minute later the beat process emits another instance of
the task. And so on.

So there's only ever one instance of each task on the queue at the same
time. Losing the contents of the queue (e.g. in a RabbitMQ crash) would
be no loss: when RabbitMQ comes back up the beat process will just
enqueue the next instance of each periodic task.

In other words h-periodic isn't really using RabbitMQ as a "queue" at
all in the normal sense of the word, it's just using it as a way to
schedule periodic tasks.

Additionally, we actively _don't want_ the queue to be durable. If
RabbitMQ goes down for a time we don't want it to come back up and the
workers to then start executing old instances of tasks. We just want
h-periodic to enqueue the _next_ instance of each periodic task and for
the workers to execute that. For example we don't want multiple workers
to be executing different instances of the same task at once.

### CODE REVIEW

**Comment by jon-betts** [comment]
I was trying to get my head around this by looking in the Flower settings, but I couldn't see anything sane. I couldn't spot any non durable queues or anything that directly lined up with the names. I think `celery` takes the name you give it and either ignores it or riffs on it and creates all sorts from it. Or... more likely I just don't understand...

Anyway, I'm not sure that making this non-durable (even though it appears to be already) is actually what we want, or at least it's not what we want 100% of the time. For things like our periodic tasks what you say is true, but I think all of our email sending is in the same queue. This is the only place I think this data is stored. So if rabbit gets bounced, all the emails in the queue would be lost.

I think (if you know what you're doing and I don't) this is actually easy enough to fix. We need to specify that the messages for certain `celery` tasks like the email one are on a separate durable queue, and the default for periodic clean-up etc stays non-durable.

These queues are also defined in `h.celery` as durable True, so I think we'd just be moving the error from `h-periodic` to `h`.

**Comment by seanh** [comment]
I'm seeing the same exception in h's logs:

```
Sep 14 15:10:39 h-prod_i-0b16fee91704b5eab eb-03ccb77de48f-stdouterr.log worker (stderr)      | amqp.exceptions.PreconditionFailed: Exchange.declare: (406) PRECONDITION_FAILED - inequivalent arg 'durable' for exchange 'celery' in vhost 'h': received 'true' but current is 'false'
```

Which you'd expect if h is also declaring the exchange with `durable=True` whereas the actual existing exchange has `durable=False`

**Comment by seanh** [comment]
> I think (if you know what you're doing and I don't) this is actually easy enough to fix. We need to specify that the messages for certain celery tasks like the email one are on a separate durable queue, and the default for periodic clean-up etc stays non-durable.

My only point of confusion is that the error messages are about a non-durable _exchange_ not a non-durable queue but I'm not sure exactly what that means when you combine Rabbit and Celery.

**Comment by seanh** [comment]
I want to do some reading up on RabbitMQ and Celery exchanges and queues to get more to the bottom of exactly what's going on here. But I think the course of action I'd recommend is:

1. Merge this PR. The `"celery"` exchange currently in RabbitMQ is non-durable and for the periodic tasks in h-periodic we actually want it to be non-durable. This PR will get rid of the warnings.

2. Merge a similar PR in h to declare the `"celery"` exchange in h as non-durable. Again, non-durable is what the exchange actually is in RabbitMQ and this will get rid of the same error message from h's logs

3. Consider moving h's email-sending tasks to a separate, durable exchange and queue. It may not be urgent to do this part though since I think they've _always_ been on a non-durable queue and it has been fine, even if it seems like they ought to be on a durable queue. The fact is that RabbitMQ / CloudAMQP doesn't crash or restart very often.

   There are other problems with sending emails too. If adding the send-email task to RabbitMQ fails (e.g. because RabbitMQ is down or is being unreliable) then you'll never get your activation or notification email. If we want to make sure that email sending is really as reliable as it can be I think it has to use the Postgres job queue as syncing annotations does. Moving it to a durable queue will make it more reliable in some ways but still not reliable in others. And if it were added to the Postgres job queue then it wouldn't _need_ to have a durable RabbitMQ queue anymore.

**Comment by marcospri** [comment]
> non-durable is what the exchange actually is in RabbitMQ and this will get rid of the same error message from h's logs

Do we now why? Is a cost saving thing on CloudAMQP?

https://www.cloudamqp.com/docs/persistence.html Looks like it's the default, I'm guessing that you might get more performance in a non durable one if you publish a lot of messages?

**Comment by seanh** [comment]
> > non-durable is what the exchange actually is in RabbitMQ and this will get rid of the same error message from h's logs
> 
> Do we now why? Is a cost saving thing on CloudAMQP?

No, no idea.

**Comment by seanh** [comment]
I've been looking into this some more. Here's a little info about "exchanges" versus "queues" in RabbitMQ:

1. Publishers (like h and h-periodic) don't publish messages onto queues. You always publish messages onto an **exchange**. The publisher actually doesn't even know what queue(s) (if any) the message will end up on. That's up to the exchange.

2. An **exchange** is a thing that accepts messages and puts them onto queues. The exchange decides which queue the message will go on, and may even place a single message onto multiple queues or discard a message without placing it onto any queues. The type of exchange we're using (a "direct exchange") places each message onto one queue.

3. In order for a message to be persistent (saved to disk, and survive RabbitMQ restarts) the message itself must be marked as "persistent" and also the queue that the message is on must be marked as "durable".

4. Exchanges can also be marked as "durable" or not. I'm not sure whether the exchange also has to be durable in order for the message to be persisted. I suspect not since message's only pass through exchanges they don't hang around on them. But I don't know.

Looking in the **RabbitMQ Manager** in our CloudAMQP account:

1. The virtual host `h` has a direct exchange named `celery`. The exchange **is durable** and it's bound to a queue named `celery` with the routing key (binding key) `celery`:

   Screenshot from 2021-09-15 12-15-09

2. The `celery` queue **is also durable**:

   Screenshot from 2021-09-15 12-17-18

The error message that we're seeing in Papertrail is `inequivalent arg 'durable' for exchange 'celery' in vhost 'h': received 'true' but current is 'false'`. This does not make sense to me: the current `durable` setting for the `celery` exchange in the vhost `h` is `true`, not `false`. See the screenshot above.

**Comment by seanh** [comment]
Slack thread: https://hypothes-is.slack.com/archives/C1MA4E9B9/p1631705052009400

**Comment by seanh** [comment]
I've sent a message to CloudAMQP support. Let's see what they say..

**Comment by seanh** [comment]
It turns out that the error messages were coming from the _Canadian_ instances of h and h-periodic. The Canadian instances use a separate RabbitMQ cluster and the `celery` exchange in that cluster must be incorrectly set to non-durable (though no one has verified this yet). The logs from the Canadian instances were being accidentally mixed in with the logs from the main instances in Papertrail so we thought the error messages were coming from the main instances.

We've now separated the logs in Papertrail and we're no longer seeing the error message from the main instances.

So the error message means exactly what it sounds like it means: h and h-periodic are both correctly trying to declare the `celery` exchange as durable but the existing `celery` exchange in RabbitMQ is set to non-durable.

Slack threads:

* https://hypothes-is.slack.com/archives/CR3E3S7K8/p1631720070007600
* https://hypothes-is.slack.com/archives/CR3E3S7K8/p1631782381031600

The remaining mystery is _how_ exactly the Canada exchange got created as non-durable? I did some testing locally:

1. Make sure you *don't* have either h or h-periodic running

2. In the `h` directory take down the services:

   ```terminal
   $ make service args=down
   ```

3. Bring them back up again:

   ```terminal
   $ make services
   ```

4. At this point there are no Celery exchanges in RabbitMQ yet, just RabbitMQ's own internal ones:

   ```terminal
   $ docker-compose exec rabbit rabbitmqctl list_exchanges
   Listing exchanges
           direct
   amq.direct      direct
   amq.topic       topic
   amq.headers     headers
   amq.rabbitmq.trace      topic
   amq.match       headers
   amq.rabbitmq.log        topic
   amq.fanout      fanout
   ```

5. Now start h and it will declare the Celery exchanges:

   ```terminal
   $ make dev
   ```

6. The exchange has been created and is durable:

   <pre><code>$ docker-compose exec rabbit rabbitmqctl list_exchanges name durable
   Listing exchanges
   celery.pidbox   false
   reply.celery.pidbox     false
   indexer true
   <b>celery  true</b>
   celeryev        true
           true
   amq.direct      true
   amq.topic       true
   amq.headers     true
   amq.rabbitmq.trace      true
   realtime        false
   amq.match       true
   amq.rabbitmq.log        true
   amq.fanout      true</code></pre>

You can repeat the same steps but run `make dev` in h-periodic instead of h and verify that if h-periodic is the first one to declare the `celery` exchange it creates it as durable too.
